File: ta4j-core/src/main/java/org/ta4j/core/analysis/cost/LinearTransactionCostModel.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.analysis.cost;

import java.util.Objects;
import org.ta4j.core.FuturesContract;
import org.ta4j.core.Position;
import org.ta4j.core.Trade;
import org.ta4j.core.TradeFill;
import org.ta4j.core.num.Num;
import org.ta4j.core.num.NumFactory;

/**
 * With this cost model, the trading costs for opening or closing a position
 * accrue linearly.
 */
public class LinearTransactionCostModel implements CostModel {

    /** The slope of the linear model (fee per position). */
    private final double feePerPosition;

    /**
     * Constructor with {@code feePerPosition * x}.
     *
     * @param feePerPosition the feePerPosition coefficient (e.g. 0.005 for 0.5% per
     *                       {@link Trade trade})
     */
    public LinearTransactionCostModel(double feePerPosition) {
        this.feePerPosition = feePerPosition;
    }

    /**
     * @param position     the position
     * @param currentIndex current bar index through which futures fills are
     *                     included
     * @return the trading cost of the single {@code position}
     */
    @Override
    public Num calculate(Position position, int currentIndex) {
        Trade entryTrade = position.getEntry();
        if (entryTrade != null && entryTrade.getFuturesContract() != null) {
            return calculateFuturesPosition(position, currentIndex);
        }
        return this.calculate(position);
    }

    @Override
    public Num calculate(Position position) {
        Trade entryTrade = position.getEntry();
        if (entryTrade != null && entryTrade.getFuturesContract() != null) {
            return calculateFuturesPosition(position, Integer.MAX_VALUE);
        }
        Num totalPositionCost = null;
        if (entryTrade != null) {
            // transaction costs of the entry trade
            totalPositionCost = entryTrade.getCost();
            if (position.getExit() != null) {
                totalPositionCost = totalPositionCost.plus(position.getExit().getCost());
            }
        }
        return totalPositionCost;
    }

    private Num calculateFuturesPosition(Position position, int currentIndex) {
        Num totalPositionCost = calculateFuturesTradeCost(position.getEntry(), currentIndex);
        Trade exitTrade = position.getExit();
        if (exitTrade != null) {
            Num exitCost = calculateFuturesTradeCost(exitTrade, currentIndex);
            totalPositionCost = totalPositionCost.plus(totalPositionCost.getNumFactory().numOf(exitCost.getDelegate()));
        }
        return totalPositionCost;
    }

    private Num calculateFuturesTradeCost(Trade trade, int currentIndex) {
        NumFactory numFactory = trade.getPricePerAsset().getNumFactory();
        Num totalTradeCost = numFactory.zero();
        for (TradeFill fill : Trade.executionFillsOf(trade)) {
            if (fill.index() < 0 || fill.index() > currentIndex) {
                continue;
            }
            Num fillCost = fill.hasRecordedFees() ? fill.fee() : calculate(fill);
            totalTradeCost = totalTradeCost.plus(numFactory.numOf(fillCost.getDelegate()));
        }
        return totalTradeCost;
    }

    @Override
    public Num calculate(Num price, Num amount) {
        return amount.getNumFactory().numOf(feePerPosition).multipliedBy(price).multipliedBy(amount);
    }

    /**
     * Applies {@link #feePerPosition} to the settlement notional of a native
     * execution fill.
     *
     * <p>
     * The linear model is a rate on the traded notional; a futures fill trades the
     * contract settlement notional, so an inverse contract is priced on
     * {@code contracts * contractSize / price} rather than on the base quantity.
     * </p>
     *
     * @param fill the execution fill
     * @return the trading cost of {@code fill}
     * @since 0.25.1
     */
    @Override
    public Num calculate(TradeFill fill) {
        Objects.requireNonNull(fill, "fill");
        FuturesContract contract = fill.futuresContract();
        if (contract == null) {
            return calculate(fill.price(), fill.amount());
        }
        Num rate = fill.price().getNumFactory().numOf(feePerPosition);
        return contract.settlementNotional(fill.amount(), fill.price()).multipliedBy(rate);
    }

    @Override
    public boolean equals(CostModel otherModel) {
        boolean equality = false;
        if (this.getClass().equals(otherModel.getClass())) {
            equality = ((LinearTransactionCostModel) otherModel).feePerPosition == this.feePerPosition;
        }
        return equality;
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/analysis/cost/LinearTransactionCostModel.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.analysis.cost

import org.ta4j.core.Position
import org.ta4j.core.Trade
import org.ta4j.core.num.*

class LinearTransactionCostModel
/**
 * Constructor. (feePerPosition * x)
 *
 * @param feePerPosition the feePerPosition coefficient (e.g. 0.005 for 0.5% per
 * [trade][Trade])
 */(
    /**
     * Slope of the linear model - fee per position
     */
    private val feePerPosition: Double
) : CostModel {
    /**
     * Calculates the transaction cost of a position.
     *
     * @param position     the position
     * @param finalIndex current bar index (irrelevant for the
     * LinearTransactionCostModel)
     * @return the absolute trade cost
     */
    override fun calculate(position: Position, finalIndex: Int): Num? {
        return this.calculate(position)
    }

    /**
     * Calculates the transaction cost of a position.
     *
     * @param position the position
     * @return the absolute trade cost
     */
    override fun calculate(position: Position): Num? {
        var totalPositionCost: Num? = null
        val entryTrade = position.entry
        if (entryTrade != null) {
            // transaction costs of entry trade
            totalPositionCost = entryTrade.cost
            if (position.exit != null) {
                totalPositionCost = totalPositionCost!!.plus(position.exit!!.cost!!)
            }
        }
        return totalPositionCost
    }

    /**
     * @param price  execution price
     * @param amount trade amount
     * @return the absolute trade transaction cost
     */
    override fun calculate(price: Num?, amount: Num?): Num {
        return amount!!.numOf(feePerPosition).times(price!!).times(amount)
    }

    /**
     * Evaluate if two models are equal
     *
     * @param other model to compare with
     */
    override fun equals(other: Any?): Boolean {
        var equality = false
        if (this.javaClass == other!!.javaClass) {
            equality = (other as LinearTransactionCostModel?)!!.feePerPosition == feePerPosition
        }
        return equality
    }

    companion object {
        private const val serialVersionUID = -8808559507754156097L
    }
}